src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,391 @@
+        {
+            // Make sure the Start and Stop can be called from different threads and don't 
+            // stomp on the other's operation. We use the _syncLock instead of the
+            // RunLoop or StreamRef because IntPtrs are value types and can't be locked
+            lock (_syncLock)
+            {
+                // Be optimistic and only stop the stream, don't clean up unless we need to.
+                StopStream();
+            }
+        }
+
+        // -----------------------------
+        // ---- PAL layer ends here ----
+        // -----------------------------
+
+        // Flags used to create the event stream
+        private const Interop.EventStream.FSEventStreamCreateFlags EventStreamFlags = (Interop.EventStream.FSEventStreamCreateFlags.kFSEventStreamCreateFlagFileEvents |
+                                                                                        Interop.EventStream.FSEventStreamCreateFlags.kFSEventStreamCreateFlagWatchRoot);
+
+        // The last directory we were told to watch. This allows us to be smart about how we create streams
+        // and to only teardown and recreate streams when we need to.
+        private string _lastWatchedDirectory = String.Empty;
+
+        // The bitmask of events that we want to send to the user
+        private Interop.EventStream.FSEventStreamEventFlags _filterFlags = 0;
+
+        // The EventStream to listen for events on
+        private FSEventStreamRef _eventStream = IntPtr.Zero;
+
+        // The background thread to use to monitor events
+        private Thread _watcherThread = null;
+
+        // A reference to the RunLoop that we can use to start or stop a Watcher
+        private CFRunLoopRef _watcherRunLoop = IntPtr.Zero;
+
+        // Since all variables are IntPtr, Strings, or change references, using this in our locks
+        private readonly object _syncLock = new object();
+
+        // Callback delegate for the EventStream events 
+        private Interop.EventStream.FSEventStreamCallback _callback = null;
+
+        // Use an event to try to prevent StartRaisingEvents from returning before the
+        // RunLoop actually begins. This will mitigate a race condition where the watcher
+        // thread hasn't completed initialization and stop is called before the RunLoop even starts.
+        private readonly AutoResetEvent _runLoopStartedEvent = new AutoResetEvent(false);
+
+        private void CreateStreamAndStartWatcher()
+        {
+            Debug.Assert(_eventStream == IntPtr.Zero);
+            Debug.Assert(_watcherRunLoop == IntPtr.Zero);
+
+            // Make sure we only do this if there is a valid directory
+            if (String.IsNullOrEmpty(_directory) == false)
+            {
+                // Get the path to watch and verify we created the CFStringRef
+                CFStringRef path = Interop.CoreFoundation.CFStringCreateWithCString(_directory);
+                if (path == IntPtr.Zero)
```

Comment:
It should be fine without since a function call like that should get inlined and end up with the same code anyway...and if not, this seems more readable and doesn't require a function call :)